neonvm/apis/neonvm/v1/virtualmachine_types.go

```diff
@@ -57,13 +57,13 @@ type VirtualMachineSpec struct {
 	// +kubebuilder:validation:Maximum=65535
 	// +kubebuilder:default:=20183
 	// +optional
-	QMP int32 `json:"qmp"`
+	QMP int32 `json:"qmp,omitempty"`
```

Comment:
QMP must be specified (these `+kubebuilde`r markers are used to set a default instead of using defaults in the webhook code). Otherwise, the controller will crash when it try do plug/unplug devices (CPU, RAM).

Reply:
Right - here's my understanding / assumptions:

1. `+kubebuilder:default:=...` affects the generated CRD only
2. CRD defaults are applied by the API server (which only does JSON/YAML)
3. CRD defaults will only be applied if the field is not present — this is why we have to set QMP from the go client, but not from a YAML file
4. the VM controller only receives objects _after_ the API server has applied any missing defaults

I may be wrong though. Is this correct?

Reply:
if you apply `vm.yaml` without `.spec.qmp` and then fetch the spec by `kubectl get vm <name> -oyaml | grep qmp` you will see `qmp: 20183` there. Not sure about the go client, but I think it should have the same spec when you call `clientset.Get()`

Reply:
> think it should have the same spec when you call `clientset.Get()`

Hm, not quite sure what you mean - can you elaborate?

Reply:
I see e2e tests are Ok (migration as well). meaning it works as expected.